### PR TITLE
add document for typescript guidelines

### PR DESCRIPTION
**Pull request by DDDDDanica**
Provide [basic guideline for typescript](https://github.com/MetaMask/contributor-docs/blob/0a2b7c426d238adefadac7cc51eee20875d25785/docs/typescript.md)

### CODE REVIEW

File: docs/typescript.md
Comment (NicolasMassart) [inline docs/typescript.md:16]: Could you provide examples for this?

File: docs/typescript.md
Comment (NicolasMassart) [inline docs/typescript.md:20]: Could you provide examples for this?

File: docs/typescript.md
Comment (NicolasMassart) [inline docs/typescript.md:383]: Could you provide examples for this?

File: docs/typescript.md
Comment (NicolasMassart) [inline docs/typescript.md:387]: Could you provide examples for this?

File: docs/typescript.md
Comment (NicolasMassart) [inline docs/typescript.md:471]: Could you provide examples for this?

**Review by NicolasMassart — COMMENTED** [review]
Nice work!
Will definitely help us.
I just pointed places where, as a new TS dev, I could be helped by some examples.

**Review by MajorLift — CHANGES_REQUESTED** [review]
Hi, great work putting this together! 

I noticed that you included my work in this (https://github.com/MetaMask/contributor-docs/issues/69#issuecomment-1910430960). These entries are still WIP, and not intended to be published yet. I'd appreciate if you could remove them. I'll submit them when they're ready.

Also, a heads-up that my PR for https://github.com/MetaMask/contributor-docs/issues/47 is almost ready for review, so any content here relating to `any`, type assertions, type annotations may be edited to remove duplication and make the doc read coherently.

File: docs/typescript.md
Comment (MajorLift) [reply docs/typescript.md:70]: This also conflicts directly with [lines 412-439](https://github.com/MetaMask/contributor-docs/pull/80#discussion_r1506095370).

File: docs/typescript.md
Comment (MajorLift) [inline docs/typescript.md:170]: There are four different entries providing conflicting guidelines for `reduce` usage. 

This highlighted portion aligns with this entry:
- https://github.com/MetaMask/contributor-docs/pull/80/files#r1506091213 (lines 38-70)

...and directly contradicts these entries:
- https://github.com/MetaMask/contributor-docs/pull/80/files#r1506093304 (lines 253-285)
- https://github.com/MetaMask/contributor-docs/pull/80/files#r1506095370 (lines 412-439)

**Comment by DDDDDanica** [comment]
Hey @MajorLift thanks for your review, when i was taking the ticket, i just collected the points in [this ticket](https://github.com/MetaMask/contributor-docs/issues/69) and didn't notice your work, will remove now

File: docs/typescript.md
Comment (MajorLift) [inline docs/typescript.md:159]: Type annotations prevent inference-based narrowing of the user-supplied type. This is why they should be avoided, or at least why type inference should generally be preferred (unless supplying the annotation narrows the type further than the inferred result).

In contrast, the `satisfies` operator enables the user to assign a wider/abstract type to a value as a constraint, while also allowing the value's type to be further narrowed.

It lets us avoid wrapping specific types in a function to assign a wider type.

> Need to create a data structure with specific values, but ensure that it's declared with a more abstract type? Create a function that takes an argument of the type you want, then use it to build the object. (https://github.com/MetaMask/contributor-docs/issues/69#issuecomment-1876114356)

```ts
const limitExceeded = () =>
    ({
      jsonrpc: '2.0',
      id: IdGenerator(),
      error: {
        code: -32005,
        message: 'Limit exceeded',
      },
    }),
```

I'll update this section with my PR for https://github.com/MetaMask/contributor-docs/issues/47, since I already cover these points there.

File: docs/typescript.md
Comment (MajorLift) [inline docs/typescript.md:9]: FYI: We're enforcing usage of type aliases over interfaces in all of our packages with the eslint rule `@typescript-eslint/consistent-type-definitions`.

See the following for previous discussions on this matter:
- https://github.com/MetaMask/eslint-config/pull/216
- https://github.com/MetaMask/core/pull/1933

File: docs/typescript.md
Comment (MajorLift) [inline docs/typescript.md:125]: The type assertion can be easily avoided here:

```suggestion
for (const key of directions) {
  const direction = directions[key];
  console.log(direction); // Output: Direction.Up, Direction.Down
}
```

File: docs/typescript.md
Comment (MajorLift) [inline docs/typescript.md:80]: Generic types are completely different from type assertions, especially since TypeScript doesn't do type casting.

See https://github.com/Microsoft/TypeScript/wiki/FAQ#why-doesnt-this-incorrect-cast-throw-a-runtime-error

```suggestion
Type assertions inform TypeScript about a variable's actual type, using the `as` syntax.
```

File: docs/typescript.md
Comment (MajorLift) [inline docs/typescript.md:39]: This would be a much better example of generic constraints with the following change:

```suggestion
// before
function createExampleMiddleware<Params, Result>(
  exampleParam,
): JsonRpcMiddleware<JsonRpcParams, Json>;
// after
function createExampleMiddleware<
  Params extends JsonRpcParams,
  Result extends Json,
>(exampleParam): JsonRpcMiddleware<Params, Result>;
```
(removed the default params to keep the example focused on generic constraints and avoid confusion)

See: https://www.typescriptlang.org/docs/handbook/2/generics.html#generic-constraints

File: docs/typescript.md
Comment (MajorLift) [inline docs/typescript.md:13]: The name for enumerated constants in TypeScript is `enum`s not "enumerations." 

```suggestion
### Enums

TypeScript offers several tools for crafting clear data definitions, with enums and unions standing as popular choices.
```

File: docs/typescript.md
Comment (MajorLift) [inline docs/typescript.md:1]:
```suggestion
# Typescript Guidelines
```
- All words should be capitalized in document title. 
- Applying bold style to a heading doesn't have any effect. All instances of this should be fixed throughout the doc.

**Review by MajorLift — COMMENTED** [review]
@DDDDDanica No worries, thanks for responding quickly! I had a few more suggestions and corrections:

File: docs/typescript.md
Comment (MajorLift) [reply docs/typescript.md:125]: Also, `for .. in` loops are banned by our eslint rules.
- For previous discussion, see https://github.com/MetaMask/core/pull/1865

File: docs/typescript.md
Comment (MajorLift) [reply docs/typescript.md:80]: Looks like this is a deprecated syntax for type assertions. In any case, it should be removed as it's no longer in use.

File: docs/typescript.md
Comment (DDDDDanica) [reply docs/typescript.md:1]: Gotcha !

File: docs/typescript.md
Comment (DDDDDanica) [reply docs/typescript.md:9]: Will change the title to `type`, thanks for the heads up

File: docs/typescript.md
Comment (DDDDDanica) [reply docs/typescript.md:17]: yes added !

File: docs/typescript.md
Comment (DDDDDanica) [reply docs/typescript.md:21]: yes, added !

File: docs/typescript.md
Comment (DDDDDanica) [reply docs/typescript.md:125]: Thanks for the heads up, applied the changes now. thanks so much !

File: docs/typescript.md
Comment (MajorLift) [inline docs/typescript.md:9]: (continued from https://github.com/MetaMask/contributor-docs/pull/80#discussion_r1506360915)

On second glance, we do use interfaces in the sense that you're describing it. And I think establishing contracts for classes is a development pattern worth encouraging. We just consistently use type alias syntax instead of interface syntax to write all types, including class "interfaces." Sorry for the confusion here. 

How about we split this into two sections?

```suggestion
## `implements` keyword

The `implements` keyword enables us to define and enforce interfaces, i.e. strict contracts consisting of expected object and class properties and abstract method signatures. 

Writing an interface to establish the specifications of a class that external code can interact with without being aware of internal implementation details is encouraged as sound OOP development practice. 

Here's an abbreviated example from `@metamask/polling-controller` of an interface being used to define one of our most important constructs.

\```ts
export type IPollingController = {
  ...
}
...
export function AbstractPollingControllerBaseMixin<TBase extends Constructor>(
  Base: TBase,
) {
  abstract class AbstractPollingControllerBase
    extends Base
    implements IPollingController
  { ... }
  return AbstractPollingControllerBase  
}
\```

The concept of the interface as discussed in this section is not to be confused with interface syntax as opposed to type alias. Note that in the above example, the `IPollingController` interface is defined as a type alias, not using the `interface` keyword. 

## Always prefer Type Aliases over Interface syntax

We enforce consistent and exclusive usage of type aliases over interface syntax for several reasons:

- The capabilities of type aliases is a strict superset of those of interfaces. 
  - Crucially, `extends`, `implements` are also supported by type aliases.
  - Declaration merging is the only exception, but we have no use case for this feature that cannot be substituted by using type intersections.
- Unlike interfaces, type aliases extend `Record` and have an index signature of `string` by default, which makes them compatible with our Json-serializable types (most notably `Record<string, Json>`). 
- Interfaces do not support multiple inheritance, while type aliases can be freely merged using the intersection (`&`) operator.
```

(remove the backticks around the fenced code block)

This is intended as an initial draft. Feel free to edit.

File: docs/typescript.md
Comment (MajorLift) [inline docs/typescript.md:165]: This is no longer an example of `as` assertion being used. Should we replace it with one that fits under the section heading?

> #### Avoid type assertions, or explain their purpose whenever necessary.

File: docs/typescript.md
Comment (MajorLift) [inline docs/typescript.md:15]: The entries in this doc should be at the same heading level.

```suggestion
#### Consider using enums over union types for situations with a fixed set of known values.
```

File: docs/typescript.md
Comment (MajorLift) [inline docs/typescript.md:35]: The entries in this doc should be at the same heading level.

```suggestion
#### Don't use numeric enums
```

File: docs/typescript.md
Comment (MajorLift) [inline docs/typescript.md:11]: Applying bold style to a markdown heading has no effect.

```suggestion
### Enums
```

File: docs/typescript.md
Comment (MajorLift) [inline docs/typescript.md:64]: Applying bold style to a markdown heading has no effect.

```suggestion
## Types
```

File: docs/typescript.md
Comment (MajorLift) [inline docs/typescript.md:5]:
```suggestion
**_Audience_**: Internal and external developers using TypeScript to contribute to MetaMask.
```

We want to use consistent styling for our main product name (similarly, "Consensys," not "ConsenSys" or some other variation).

**Review by MajorLift — COMMENTED** [review]
Thanks for applying the suggestions! Some more notes:

File: docs/typescript.md
Comment (MajorLift) [reply docs/typescript.md]: Continued in https://github.com/MetaMask/contributor-docs/pull/80#discussion_r1506955448

File: docs/typescript.md
Comment (MajorLift) [reply docs/typescript.md]: Continued in https://github.com/MetaMask/contributor-docs/pull/80#discussion_r1506956360

File: docs/typescript.md
Comment (MajorLift) [reply docs/typescript.md:9]: Removing the second entry is also acceptable since `@typescript-eslint/consistent-type-definitions` does a good job of enforcing the practice without everyone having to know the reasoning behind its adoption. 

In general, the guidelines doesn't need to cover conventions that we already enforce using linter rules or static code analysis in too much detail. This doc isn't intended as a comprehensive collection of every preference we have agreed upon. The priority should be on covering principles or processes that aren't reducible to simple rulesets and rely on a deeper understanding from the developers to be effective.

File: docs/typescript.md
Comment (MajorLift) [reply docs/typescript.md:165]: Ah actually this is my bad. I see this example is working by coincidence: because `Direction.Up` and `Direction.Down` happen to be the first two index keys of the `directions` array.

This would still work as an example of type assertion with the following changes:

```typescript
enum Direction {
  Up = 'up',
  Down = 'down',
  Left = 'left',
  Right = 'right',
}

const directions: Direction[] = [Direction.Up, Direction.Down];

for (const key of directions) {
  const direction = directions[key as keyof typeof directions];
  console.log(direction); // Output: Direction.Up, Direction.Down
}
```

We also have a built-in utility type for narrowing key signatures, which simplifies the code.

```ts
import { getKnownPropertyNames } from '@metamask/utils';

...

for (const key of getKnonwPropertyNames(direction)) {
  const direction = directions[key];
}
```

File: docs/typescript.md
Comment (DDDDDanica) [reply docs/typescript.md:9]: Thanks for all the recommendations, i've added your suggestions, and changed the section title as `Implements`

File: docs/typescript.md
Comment (DDDDDanica) [reply docs/typescript.md:15]: I'm using `**` to represent lower level title, but can certainly change everyone !

File: docs/typescript.md
Comment (DDDDDanica) [reply docs/typescript.md:64]: changed also

File: docs/typescript.md
Comment (DDDDDanica) [reply docs/typescript.md:165]: Thanks for the suggestion, i've learnt a new`getKnonwPropertyNames` today

File: docs/typescript.md
Comment (MajorLift) [inline docs/typescript.md:193]: I should have clarified that my `getKnownPropertyNames` example wasn't meant to be used here. I'm sorry for the confusion.

The usefulness of `getKnownPropertyNames` is that it makes the `as` assertion *unnecessary* ([you can see why by checking out its definition](https://github.com/MetaMask/utils/blob/ec0237cf6793b9b31efffde247152c12ffd1a260/src/misc.ts#L102-L117)). So the above code can't be used as an example for this section.

We'll need to remove the `getKnownPropertyNames` import statement. The type annotation for `directions` can also be removed (this is preferable for reasons detailed [here](https://github.com/MetaMask/contributor-docs/pull/74/files#diff-ae03a9f2eb638a25c1d8d7ba853c78220f1ed44a19ad97a8d780cbc0c00a4260R7-R46)).

```ts
enum Direction { 
  Up = 'up', 
  Down = 'down', 
  Left = 'left', 
  Right = 'right', 
}
const directions = [Direction.Up, Direction.Down];
```

From there, we can resolve the index signature error that arises from the property access expression `directions[key]`,
> Element implicitly has an 'any' type because index expression is not of type 'number'.(7015)

by choosing one of these two options:

- option 1:
```ts
for (const key of Object.keys(direction)) {
  const direction = directions[key as keyof typeof directions];
}
```
- option 2:
```ts
for (const key of Object.keys(directions) as (keyof typeof directions)[]) {
  const direction = directions[key];
}
```

You can use either or both as the example(s) for this entry.

File: docs/typescript.md
Comment (MajorLift) [inline docs/typescript.md:9]: Nit: is this maybe a section heading rather than just emphasized text?

File: docs/typescript.md
Comment (MajorLift) [inline docs/typescript.md:32]: Nit: is this maybe a section heading rather than just emphasized text?

File: docs/typescript.md
Comment (MajorLift) [inline docs/typescript.md:45]: Nit:
> MD018/no-missing-space-atx: No space after hash on atx style heading markdownlint [MD018](https://github.com/DavidAnson/markdownlint/blob/v0.33.0/doc/md018.md)

```suggestion
#### Consider using enums over union types for situations with a fixed set of known values.
```

File: docs/typescript.md
Comment (MajorLift) [inline docs/typescript.md:65]: Nit:
> MD018/no-missing-space-atx: No space after hash on atx style heading markdownlint [MD018](https://github.com/DavidAnson/markdownlint/blob/v0.33.0/doc/md018.md)

```suggestion
#### Don't use numeric enums
```

**Review by MajorLift — COMMENTED** [review]
Looks like we're almost there! Just one important fix and some linting issues left.

On the subject of linting, I recommend setting up the [`markdownlint` extension](https://marketplace.visualstudio.com/items?itemName=DavidAnson.vscode-markdownlint) in your IDE. Its ruleset is mostly consistent with our prettier config and a lot more comprehensive.

Having a shared set of conventions on how to structure and format these docs helps future collaborators add their contributions without feeling like they're stepping on any toes.

On why consistency in the heading hierarchy is a good thing, check out the outline view in the preview for this doc: https://github.com/MetaMask/contributor-docs/blob/feature/69-typescript/docs/typescript.md. You can enable it by clicking on the unordered-list icon on the top right (`Button.types__StyledButton-sc-ws60qy-0`).

I think we should be ready to merge after this next round of fixes. Thanks for bearing with me through all these comments.

File: docs/typescript.md
Comment (DDDDDanica) [reply docs/typescript.md:32]: Good catch !! changed !

File: docs/typescript.md
Comment (DDDDDanica) [reply docs/typescript.md:193]: Thank you @MajorLift for the comprehensive explanation, modified this section as well

**Review by MajorLift — DISMISSED** [review]
LGTM! 

Added a commit https://github.com/MetaMask/contributor-docs/pull/80/commits/5d990cd909b31ac6dfa3380ec73b9201a8c261dc before leaving the approval so you could merge without more back and forth. 

Feel free to ping me for another approval if there are more changes you want to add.

**Review by MajorLift — APPROVED** [review]

**Comment by DDDDDanica** [comment]
@MajorLift thanks for your review and approval, they are really helpful ! I'll draft another PR to address the rest

**Comment by DDDDDanica** [comment]
Have a conversation with @MajorLift offline to merge his work first, then add more for <TODO> that he will leave for me

**Review by bdstar81 — APPROVED** [review]

**Pull request merged**
